.github/workflows/ci.yml

```diff
@@ -68,10 +93,95 @@ jobs:
         run: |-
           mvn clean test --no-transfer-progress -f client-lib/java
 
-  # To be removed entirely when TODO(#158) is completed.
+  lint_and_unit:
+    runs-on: 'ubuntu-latest'
+    needs:
+      - 'go_lint'
+      - 'java_lint'
+      - 'terraform_lint'
+      - 'go_test'
+      - 'java_test'
+      - 'yaml_lint'
+    steps:
+      - run: 'echo prechecks complete'
+
+  # Build pmap services and push to artifact registry
+  build:
+    runs-on: 'ubuntu-latest'
+    permissions:
+      contents: 'read'
+      id-token: 'write'
+    needs: ['lint_and_unit']
+    steps:
+      - name: 'Checkout'
+        uses: 'actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c' # ratchet:actions/checkout@v3
+      - name: 'Setup Go'
+        uses: 'actions/setup-go@6edd4406fa81c3da01a34fa6f6343087c207a568' # ratchet:actions/setup-go@v3
+        with:
+          go-version: '1.21'
+      - id: 'auth'
+        name: 'Authenticate to Google Cloud'
+        uses: 'google-github-actions/auth@ef5d53e30bbcd8d0836f4288f5e50ff3e086997d' # ratchet:google-github-actions/auth@v1
+        with:
+          workload_identity_provider: '${{ env.WIF_PROVIDER }}'
+          service_account: '${{ env.WIF_SERVICE_ACCOUNT }}'
+          token_format: 'access_token'
+      - name: 'Authenticate to Artifact Registry'
+        uses: 'docker/login-action@f4ef78c080cd8ba55a85445d5b36e214a81df20a' # ratchet:docker/login-action@v2
+        with:
+          username: 'oauth2accesstoken'
+          password: '${{ steps.auth.outputs.access_token }}'
+          registry: '${{ env.CONTAINER_REGISTRY }}'
+      # goreleaser requires a tag to publish images to container registry.
+      # We create a local tag to make it happy.
+      - run: |-
+          git config user.name "${GITHUB_ACTOR}"
+          git config user.email "${GITHUB_ACTOR}@users.noreply.github.com"
+          git tag -f `date "+%Y%m%d%H%M%S"`
+      - name: 'Build the service containers and push to the registry with goreleaser'
+        uses: 'goreleaser/goreleaser-action@f82d6c1c344bcacabba2c841718984797f664a6b' # ratchet:goreleaser/goreleaser-action@v4
+        with:
+          version: 'v1.16.1' # Manually pinned
+          args: 'release -f .goreleaser.docker.yaml --clean'
+
+  deployment:
+    runs-on: 'ubuntu-latest'
+    needs:
+      - 'build'
+    permissions:
+      contents: 'read'
+      id-token: 'write'
+    strategy:
+      matrix:
+        services: ['jvs-api-4527', 'jvs-cert-rotator-e04a', 'jvs-public-key-f22e', 'jvs-ui-8ab7']
```

Comment:
Matrix doesn't support ${{ env.XXX }} so I have to hard code it here.
I think this is better than having four steps and each of them updates one services
```
- name: 'Update API service'
        run: gcloud run services update api_service
- name: 'Update cert rotator service'
        run: gcloud run services update cert_rotator
......
```

Reply:
We cannot use matrix as we might need to add kms key as env var for these services, and they use different names for it.